### PR TITLE
App Secrets, Attempt no.3

### PR DESCRIPTION
**Pull request by esreli**
This PR attempts to solve https://github.com/Esri/data-collection-ios/issues/229 once and for all. The work here builds off #244 and #246.

This attempt uses a similar technique found in the previous attempt (#246) but instead of using Apple Swift's gyb, an in-house bash script was written and incorporated in a build rule to find secrets keys and swap them with values.

Some areas to pay extra attention to in this PR include:

- [README, grant permission](https://github.com/Esri/data-collection-ios/blob/eli/app-secrets-bash/README.md#grant-permission)
- [README, configure the app](https://github.com/Esri/data-collection-ios/blob/eli/app-secrets-bash/README.md#configure-the-app)
- [Docs, app secrets section](https://github.com/Esri/data-collection-ios/tree/eli/app-secrets-bash/docs#app-secrets)
- the [script](https://github.com/Esri/data-collection-ios/blob/eli/app-secrets-bash/data-collection/scripts/masquerade)
- the [build rule](https://github.com/Esri/data-collection-ios/blob/eli/app-secrets-bash/data-collection/data-collection.xcodeproj/project.pbxproj#L204-L219)
- the [input file](https://github.com/Esri/data-collection-ios/blob/eli/app-secrets-bash/data-collection/data-collection/AppSecrets.swift.masque)

The bash script (`masquerade`) has evolved with feedback from various Esri dev-ops engineers.

> Note, this PR's base branch is currently `patch-1.2.1`. This is because this branch required branching from a hotfix branch that is not yet merged to master. Please consider this work hotfix v1.2.2 and the eventual base branch changed to `master` once `patch-1.2.1` is merged.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -40,93 +43,166 @@
...
+
+3. _Optionally_ add your **License Key** to the secrets file. Licensing the app will remove the 'Licensed for Development' watermark. Licensing the app is optional in development but required for production. Aquire your license key from your [dashboard](https://developers.arcgis.com/dashboard).
+
+  ```bash
+  echo ARCGIS_LICENSE_KEY=your-license-key >> data-collection/.secrets
+  ```
+
+  > Replace 'your-license-key' with your **License Key**.
+
+A Lite license is free and provides developers a good starting point. Visit the developers website to learn more about [licensing your ArcGIS Runtime app](https://developers.arcgis.com/pricing/licensing/).
```
Comment (jf990) [inline README.md:196]:
```suggestion
A Lite license is free and enables all of the capabilities required by this app. Visit the developer's website to learn more about [licensing your ArcGIS Runtime app](https://developers.arcgis.com/pricing/licensing/).
```

File: docs/README.md [hunk 2]
```diff
@@ -1172,6 +1176,76 @@ The `AppConfiguration` contains a series of static configuration resources. Modi
...
+
+The program looks for keys using a regex pattern `{{ *$KEY *}}`. The pattern can be chopped up into a number of sub-components.
+
+- `{{`, two open-brackets specify the start of the pattern
+- ` *`, a space-astrix permits 0-to-n space characters
+- `$KEY`, a key variable specified in the secrets file
+- ` *`, a space-astrix again permits 0-to-n space characters
+- `}}`, two close-brackets specify the end of the pattern
+
+Masquerade demands various parameters.
```
Comment (jf990) [inline docs/README.md:1195]:
```suggestion
Masquerade requires the following parameters:
```

**Comment by esreli** [comment]
> Once we have both those components in place (keep out of repo + obfuscate in binary), we'll have a "best practice" solution. Even then it won't be watertight, but it'll be really good.
@nixta 

I do agree with you here. I haven't lost sight of further obfuscation and would like to tackle that next, but wanted to solve the secrets in source problem first.

**Review by mhdostal — APPROVED** [review]
Looks excellent.  Nice work!

**Comment by esreli** [comment]
Thanks, all. I'll be merging this work into `v.next-1.2.2`. Please note, once this work is merged, contributors will need to setup their environment as described in the README. It's a simple and quick change but you will not be able to build the app without doing this step- cc @mhdostal.

**Pull request merged**